File: dSoak-Common-Java/src/MessageTester/GameDataTester.java
package MessageTester;

import static org.junit.Assert.*;
import java.io.IOException;
import java.util.ArrayList;
import org.junit.Test;
import Messages.GameData;
import Messages.Message;
import SharedObject.GameInfo;
import SharedObject.MessageNumber;
import SharedObject.ProcessData;
import SharedObject.ProcessData.PossibleProcessType;

public class GameDataTester
{
	@Test
	public void test_EveryThing() throws ClassNotFoundException, IOException
	{
		MessageNumber.LocalProcessId = 100;
		
		GameData msg1 = new GameData();
		
		assertNotNull(msg1.MessageNr);
		assertEquals(100, msg1.MessageNr.ProcessId);
		assertTrue(msg1.MessageNr.SeqNumber > 0);
		assertEquals(msg1.MessageNr, msg1.ConvId);
	
		GameInfo gameInfo = new GameInfo();
		gameInfo.GameId = 102;
		gameInfo.Status = GameInfo.StatusCode.NOTINITIALIZED;
		
		ArrayList<ProcessData> processes = new  ArrayList<ProcessData>();
		ProcessData process1 = new ProcessData();
		process1.GameId = 102;
		process1.ProcessId = 10;
		process1.ProcessType = PossibleProcessType.Player;
		
		ProcessData process2 = new ProcessData();
		process2.GameId = 102;
		process2.ProcessId = 20;
		process2.ProcessType = PossibleProcessType.Player;
		
		ProcessData process3 = new ProcessData();
		process3.GameId = 102;
		process3.ProcessId = 30;
		process3.ProcessType = PossibleProcessType.Thief;
		
		
		processes.add(process1);
		processes.add(process2);
		processes.add(process3);
		
		GameData msg2 = new GameData();
		msg2.Info = gameInfo;
		msg2.Processes  = processes;
			
		assertNotNull(msg2.MessageNr);
		assertEquals(100, msg2.MessageNr.ProcessId);
		assertEquals(msg1.MessageNr.SeqNumber + 1 , msg2.MessageNr.SeqNumber);
		assertEquals(msg2.MessageNr, msg2.ConvId);
		assertNotNull(msg2.Processes);
		assertSame(processes, msg2.Processes);
		
		byte[] bytes = msg2.Encode();
		String str = new String(bytes);
		
		Message msg3 = Message.Decode(bytes);
		assertTrue(msg3 instanceof GameData);
		GameData msg4 = (GameData) msg3;
		
		assertEquals(msg2.MessageNr.ProcessId, msg4.MessageNr.ProcessId);
		assertEquals(msg2.MessageNr.SeqNumber, msg4.MessageNr.SeqNumber);
		assertEquals(msg2.ConvId.ProcessId, msg4.ConvId.ProcessId);
		assertEquals(msg2.ConvId.SeqNumber, msg4.ConvId.SeqNumber);
		assertNotNull(msg4.Info);
		assertEquals(msg2.Info.GameId, msg4.Info.GameId);
		assertEquals(msg2.Info.Status.getValue(), msg4.Info.Status.getValue());
		assertNotNull(msg4.Processes);
		assertEquals(msg2.Processes.size(), msg4.Processes.size());
	}

}

File: dSoak-Common-CSharp/MessageTesting/GameDataTester.cs
using System;
using System.Collections.Generic;
using System.Text;
using Microsoft.VisualStudio.TestTools.UnitTesting;

using Messages;
using SharedObjects;

namespace MessageTesting
{
    [TestClass]
    public class GameDataTester
    {
        [TestInitialize]
        public void Setup()
        {
            MessageNumber.LocalProcessId = 100;
        }

        [TestMethod]
        public void GameData_CheckEverything()
        {
            GameData msg1 = new GameData();
            Assert.IsNotNull(msg1.MessageNr);
            Assert.AreEqual(100, msg1.MessageNr.ProcessId);
            Assert.IsTrue(msg1.MessageNr.SeqNumber > 0);
            Assert.AreEqual(msg1.MessageNr, msg1.ConvId);

            GameInfo gameInfo = new GameInfo() { GameId = 102, Status = GameInfo.StatusCode.NotInitialized };
            List<ProcessData> processes = new List<ProcessData>()
                {
                    new ProcessData() { GameId = 102, ProcessId = 10, ProcessType = ProcessData.PossibleProcessType.Player },
                    new ProcessData() { GameId = 102, ProcessId = 20, ProcessType = ProcessData.PossibleProcessType.Player },
                    new ProcessData() { GameId = 102, ProcessId = 30, ProcessType = ProcessData.PossibleProcessType.Thief },
                };
            GameData msg2 = new GameData() { Info = gameInfo, Processes = processes };
            Assert.IsNotNull(msg2.MessageNr);
            Assert.AreEqual(100, msg2.MessageNr.ProcessId);
            Assert.AreEqual(msg1.MessageNr.SeqNumber + 1, msg2.MessageNr.SeqNumber);
            Assert.AreEqual(msg2.MessageNr, msg2.ConvId);
            Assert.IsNotNull(msg2.Info);
            Assert.AreSame(gameInfo, msg2.Info);
            Assert.IsNotNull(msg2.Processes);
            Assert.AreSame(processes, msg2.Processes);

            byte[] bytes = msg2.Encode();
            string tmp = Encoding.ASCII.GetString(bytes);

            Message msg3 = Message.Decode(bytes);
            Assert.IsTrue(msg3 is GameData);
            GameData msg4 = msg3 as GameData;
            Assert.AreEqual(msg2.MessageNr, msg4.MessageNr);
            Assert.AreEqual(msg2.ConvId, msg4.ConvId);
            Assert.IsNotNull(msg4.Info);
            Assert.AreEqual(msg2.Info.GameId, msg4.Info.GameId);
            Assert.AreEqual(msg2.Info.Status, msg4.Info.Status);
            Assert.IsNotNull(msg4.Processes);
            Assert.AreEqual(msg2.Processes.Count, msg4.Processes.Count);
            for (int i = 0; i < msg4.Processes.Count; i++)
            {
                Assert.AreEqual(msg2.Processes[i].GameId, msg4.Processes[i].GameId);
                Assert.AreEqual(msg2.Processes[i].ProcessId, msg4.Processes[i].ProcessId);
                Assert.AreEqual(msg2.Processes[i].ProcessType, msg4.Processes[i].ProcessType);
                Assert.AreEqual(msg2.Processes[i].LifePoints, msg4.Processes[i].LifePoints);
                Assert.AreEqual(msg2.Processes[i].HitPoints, msg4.Processes[i].HitPoints);
                Assert.AreEqual(msg2.Processes[i].NumberOfPennies, msg4.Processes[i].NumberOfPennies);
                Assert.AreEqual(msg2.Processes[i].NumberOfUnfilledBalloon, msg4.Processes[i].NumberOfUnfilledBalloon);
                Assert.AreEqual(msg2.Processes[i].NumberOfFilledBalloon, msg4.Processes[i].NumberOfFilledBalloon);
                Assert.AreEqual(msg2.Processes[i].NumberOfUnraisedUmbrellas, msg4.Processes[i].NumberOfUnraisedUmbrellas);
                Assert.AreEqual(msg2.Processes[i].HasUmbrellaRaised, msg4.Processes[i].HasUmbrellaRaised);
            }
        }
    }
}
